internal/pkg/daemon/bpfrecorder/bpfrecorder.go

```diff
@@ -0,0 +1,578 @@
+	}
+
+	return nil
+}
+
+// Dial can be used to connect to the default GRPC server by creating a new
+// client.
+func Dial() (*grpc.ClientConn, context.CancelFunc, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	conn, err := grpc.DialContext(ctx, addr(), grpc.WithInsecure())
+	if err != nil {
+		cancel()
+		return nil, nil, errors.Wrap(err, "GRPC dial")
+	}
+	return conn, cancel, nil
+}
+
+// addr returns the default server listening address.
+func addr() string {
+	return net.JoinHostPort("localhost", "9112")
+}
+
+func (b *BpfRecorder) Start(
+	ctx context.Context, r *api.EmptyRequest,
+) (*api.EmptyResponse, error) {
+	if b.startRequests == 0 {
+		b.logger.Info("Starting bpf recorder")
+		if err := b.load(); err != nil {
+			return nil, errors.Wrap(err, "load bpf")
+		}
+	} else {
+		b.logger.Info("bpf recorder already running")
+	}
+
+	atomic.AddInt64(&b.startRequests, 1)
+	return &api.EmptyResponse{}, nil
+}
+
+func (b *BpfRecorder) Stop(
+	ctx context.Context, r *api.EmptyRequest,
+) (*api.EmptyResponse, error) {
+	if b.startRequests == 0 {
+		b.logger.Info("bpf recorder not running")
+		return &api.EmptyResponse{}, nil
+	}
+
+	atomic.AddInt64(&b.startRequests, -1)
+	if b.startRequests == 0 {
+		b.logger.Info("Stopping bpf recorder")
+		b.unload()
+	} else {
+		b.logger.Info("Not stopping because another recording is in progress")
+	}
+	return &api.EmptyResponse{}, nil
+}
+
+// SyscallsForNamespace returns the syscall names for the provided PID.
+func (b *BpfRecorder) SyscallsForProfile(
+	ctx context.Context, r *api.ProfileRequest,
+) (*api.SyscallsResponse, error) {
+	if b.startRequests == 0 {
+		return nil, errors.New("bpf recorder not running")
+	}
+	b.logger.Info("Getting syscalls for profile " + r.Name)
+
+	res, exist := b.pidsForProfiles.LoadAndDelete(r.Name)
+	if !exist {
+		return nil, errors.Errorf("no PID found for container")
+	}
+	pids := res.([]Pid)
+	b.logger.Info(fmt.Sprintf("Got PIDs for the profile: %v", pids))
+	if len(pids) == 0 {
+		return nil, errors.Errorf("PID slice is empty")
+	}
+
+	// We're just choosing the first PID because their mount namespace should
+	// be the same (last famous words).
+	mntns, err := b.mountNamespaceForPID(pids[0].id)
+	if err != nil {
+		return nil, errors.Wrapf(err, "get mount namespace")
+	}
+
+	mntnsKey := unsafe.Pointer(&mntns)
+	syscalls, err := b.syscalls.GetValue(mntnsKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "no syscalls found for namespace")
+	}
+
+	resp := &api.SyscallsResponse{}
+	prctlSeen := false
+	for id, set := range syscalls {
+		if set == 1 {
+			name, err := b.syscallNameForID(id)
+			if err != nil {
+				b.logger.Error(err, "unable to convert syscall ID")
+				continue
+			}
+
+			// We only start recording the syscalls from the call to prctl
+			// within the OCI runtime.
+			if name == "prctl" {
+				prctlSeen = true
+			}
+
+			if prctlSeen {
+				resp.Syscalls = append(resp.Syscalls, name)
+			}
```

Comment:
I thought about skipping the runtime (runc) completely but I think this would not work as well. Do you have anything in mind how we could determine the start of a recording in a better way?